src/libdrakvuf/drakvuf.c

```diff
@@ -1085,6 +1085,75 @@ bool drakvuf_set_vcpu_gprs(drakvuf_t drakvuf, unsigned int vcpu, registers_t* re
     return xen_set_vcpu_ctx(drakvuf->xen, drakvuf->domID, vcpu, &ctx);
 }
 
+bool drakvuf_vmi_response_set_registers(
+    drakvuf_t drakvuf,
+    drakvuf_trap_info_t* info,
+    x86_registers_t* regs,
+    bool immediate)
+{
+    bool res;
+
+    drakvuf_lock_and_get_vmi(drakvuf);
+    if (drakvuf->vmi_response_set_registers[info->vcpu])
+    {
+        PRINT_DEBUG("[LIBDRAKVUF] [%ld] Failed to set registers\n", info->event_uid);
+        res = false;
+    }
+    else
+    {
+        drakvuf->vmi_response_set_registers[info->vcpu] = true;
+        memcpy(&info->regs_modified, regs, sizeof(x86_registers_t));
+        if (immediate)
+            memcpy(info->regs, regs, sizeof(x86_registers_t));
+        res = true;
+    }
+    drakvuf_release_vmi(drakvuf);
+
+    return res;
+}
+
+bool drakvuf_is_active_callback(drakvuf_t drakvuf, drakvuf_trap_info_t* info)
+{
+    drakvuf_lock_and_get_vmi(drakvuf);
+    void* injection_cb = drakvuf_lookup_injection(drakvuf, info);
+    bool res = !injection_cb || injection_cb == info->trap->cb;
+    drakvuf_release_vmi(drakvuf);
+    return res;
+}
+
+void* drakvuf_lookup_injection(drakvuf_t drakvuf, drakvuf_trap_info_t* info)
+{
+    drakvuf_lock_and_get_vmi(drakvuf);
+    uint64_t pid = info->attached_proc_data.pid;
+    uint64_t tid = info->attached_proc_data.tid;
+    gpointer key = (gpointer)(pid << 32 | tid);
+    gpointer value = g_hash_table_lookup(drakvuf->injections_in_progress, key);
+    drakvuf_release_vmi(drakvuf);
+    return value;
+}
+
+void drakvuf_insert_injection(drakvuf_t drakvuf,
+    drakvuf_trap_info_t* info,
+    event_response_t (*cb)(drakvuf_t, drakvuf_trap_info_t*))
```

Comment:
Any particular reason why the callback is needed? Doesn't seem to be used later

Reply:
@tklengyel is this OK to you?